File: riveredge-frontend/src/apps/kuaizhizao/services/sales-review.ts
/**
 * 订单评审 API
 */

import { apiRequest } from '../../../services/api';

export type SalesReviewStatus =
  | 'draft'
  | 'reviewing'
  | 'rejected'
  | 'passed'
  | 'closed'
  | 'cancelled';

export type SalesReviewDeptCode = 'tech' | 'process' | 'purchase' | 'production' | 'quality';

export interface SalesReviewItem {
  id?: number;
  sales_review_id?: number;
  line_no?: number;
  material_id?: number | null;
  material_code: string;
  material_name: string;
  material_spec?: string | null;
  material_unit?: string | null;
  quantity: number | string;
  unit_price?: number | string;
  amount?: number | string | null;
  tech_requirements?: string | null;
  notes?: string | null;
}

export interface SalesReviewDeptOpinion {
  id: number;
  sales_review_id: number;
  review_round: number;
  dept_code: string;
  result: string;
  opinion?: string | null;
  reviewed_by?: number | null;
  reviewed_by_name?: string | null;
  reviewed_at?: string | null;
}

export interface SalesReview {
  id: number;
  uuid?: string;
  tenant_id?: number;
  review_code: string;
  status: SalesReviewStatus | string;
  review_round: number;
  customer_id: number;
  customer_code?: string | null;
  customer_name: string;
  customer_contact?: string | null;
  customer_phone?: string | null;
  project_name: string;
  review_date?: string | null;
  delivery_date?: string | null;
  urgency: string;
  risk_level: string;
  settlement_method?: string | null;
  payment_cycle?: string | null;
  delivery_location?: string | null;
  transport_method?: string | null;
  material_desc?: string | null;
  spec_desc?: string | null;
  process_desc?: string | null;
  packaging_req?: string | null;
  production_notes?: string | null;
  sales_opinion?: string | null;
  final_conclusion?: string | null;
  remarks?: string | null;
  attachments?: unknown[] | null;
  quotation_id?: number | null;
  quotation_code?: string | null;
  customer_follow_up_id?: number | null;
  salesman_id?: number | null;
  salesman_name?: string | null;
  sales_order_id?: number | null;
  sales_order_code?: string | null;
  total_quantity?: number | string;
  total_amount?: number | string;
  items?: SalesReviewItem[];
  dept_opinions?: SalesReviewDeptOpinion[];
  created_at?: string;
  updated_at?: string;
  created_by?: number | null;
  updated_by?: number | null;
  created_by_name?: string | null;
  updated_by_name?: string | null;
}

export interface SalesReviewListItem {
  id: number;
  review_code: string;
  customer_id: number;
  customer_name: string;
  project_name: string;
  status: string;
  review_round: number;
  urgency: string;
  risk_level: string;
  delivery_date?: string | null;
  review_date?: string | null;
  total_quantity?: number | string;
  total_amount?: number | string;
  salesman_name?: string | null;
  sales_order_code?: string | null;
  created_at?: string;
  updated_at?: string;
  created_by?: number | null;
  created_by_name?: string | null;
  updated_by?: number | null;
  updated_by_name?: string | null;
}

export interface SalesReviewListResult {
  items: SalesReviewListItem[];
  total: number;
  skip?: number;
  limit?: number;
}

export interface SalesReviewListParams {
  skip?: number;
  limit?: number;
  status?: string;
  customer_id?: number;
  keyword?: string;
  order_by?: string;
  /** 仅可下推销售订单：评审已通过且未关联销售订单 */
  pullable_only?: boolean;
}

export interface SalesReviewItemInput {
  material_id?: number | null;
  material_code: string;
  material_name: string;
  material_spec?: string | null;
  material_unit?: string | null;
  quantity: number | string;
  unit_price?: number | string;
  tech_requirements?: string | null;
  notes?: string | null;
}

export interface SalesReviewCreatePayload {
  review_code?: string | null;
  customer_id: number;
  customer_code?: string | null;
  customer_name: string;
  customer_contact?: string | null;
  customer_phone?: string | null;
  project_name: string;
  review_date?: string | null;
  delivery_date?: string | null;
  urgency?: string;
  risk_level?: string;
  settlement_method?: string | null;
  payment_cycle?: string | null;
  delivery_location?: string | null;
  transport_method?: string | null;
  material_desc?: string | null;
  spec_desc?: string | null;
  process_desc?: string | null;
  packaging_req?: string | null;
  production_notes?: string | null;
  sales_opinion?: string | null;
  remarks?: string | null;
  quotation_id?: number | null;
  quotation_code?: string | null;
  customer_follow_up_id?: number | null;
  salesman_id?: number | null;
  salesman_name?: string | null;
  items: SalesReviewItemInput[];
}

export type SalesReviewUpdatePayload = Partial<Omit<SalesReviewCreatePayload, 'items'>> & {
  items?: SalesReviewItemInput[];
  final_conclusion?: string | null;
};

export interface SalesReviewDeptOpinionSubmit {
  result: 'pass' | 'fail';
  opinion?: string | null;
}

export interface SalesReviewPushPreview {
  can_push: boolean;
  blocking_reason?: string | null;
  review_code: string;
  customer_name: string;
  item_count: number;
  total_quantity: number | string;
  total_amount: number | string;
  items: Record<string, unknown>[];
}

export interface SalesReviewPushResult {
  success: boolean;
  message: string;
  sales_order_id?: number | null;
  sales_order_code?: string | null;
}

const BASE = '/apps/kuaizhizao/sales-reviews';

export const salesReviewApi = {
  list: async (params?: SalesReviewListParams): Promise<SalesReviewListResult> =>
    apiRequest(BASE, { method: 'GET', params }),

  get: async (id: number): Promise<SalesReview> =>
    apiRequest(`${BASE}/${id}`, { method: 'GET' }),

  create: async (data: SalesReviewCreatePayload): Promise<SalesReview> =>
    apiRequest(BASE, { method: 'POST', data }),

  update: async (id: number, data: SalesReviewUpdatePayload): Promise<SalesReview> =>
    apiRequest(`${BASE}/${id}`, { method: 'PUT', data }),

  remove: async (id: number): Promise<void> =>
    apiRequest(`${BASE}/${id}`, { method: 'DELETE' }),

  issue: async (id: number): Promise<SalesReview> =>
    apiRequest(`${BASE}/${id}/issue`, { method: 'POST' }),

  withdraw: async (id: number): Promise<SalesReview> =>
    apiRequest(`${BASE}/${id}/withdraw`, { method: 'POST' }),

  submitDeptOpinion: async (
    id: number,
    deptCode: string,
    data: SalesReviewDeptOpinionSubmit,
  ): Promise<SalesReview> =>
    apiRequest(`${BASE}/${id}/dept-opinions/${deptCode}`, { method: 'POST', data }),

  reject: async (id: number, reason?: string | null): Promise<SalesReview> =>
    apiRequest(`${BASE}/${id}/reject`, {
      method: 'POST',
      data: reason != null && String(reason).trim() ? { reason: String(reason).trim() } : {},
    }),

  previewPushToSalesOrder: async (id: number): Promise<SalesReviewPushPreview> =>
    apiRequest(`${BASE}/${id}/push-to-sales-order/preview`, { method: 'GET' }),

  pushToSalesOrder: async (id: number): Promise<SalesReviewPushResult> =>
    apiRequest(`${BASE}/${id}/push-to-sales-order`, { method: 'POST' }),

  pullFromQuotation: async (quotationId: number): Promise<{
    success: boolean;
    message?: string;
    sales_review?: SalesReview;
  }> =>
    apiRequest(`${BASE}/pull-from-quotation`, {
      method: 'POST',
      data: { quotation_id: quotationId },
    }),
};

File: riveredge-backend/src/apps/kuaizhizao/schemas/sales_review.py
"""订单评审 Schema"""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Any, Dict, List, Optional

from pydantic import Field, model_validator

from core.schemas.base import BaseSchema


SALES_REVIEW_DEPT_CODES = ("tech", "process", "purchase", "production", "quality")


class SalesReviewItemBase(BaseSchema):
    material_id: Optional[int] = Field(None, description="物料ID")
    material_code: str = Field(..., max_length=50, description="物料编码")
    material_name: str = Field(..., max_length=200, description="物料名称")
    material_spec: Optional[str] = Field(None, max_length=200, description="规格")
    material_unit: Optional[str] = Field(None, max_length=20, description="单位")
    quantity: Decimal = Field(..., gt=0, description="数量")
    unit_price: Decimal = Field(Decimal("0"), ge=0, description="单价")
    amount: Optional[Decimal] = Field(None, description="金额（服务端重算）")
    tech_requirements: Optional[str] = Field(None, description="技术要求")
    notes: Optional[str] = Field(None, description="备注")


class SalesReviewItemCreate(SalesReviewItemBase):
    pass


class SalesReviewItemUpdate(SalesReviewItemBase):
    id: Optional[int] = Field(None, description="明细ID（更新时）")


class SalesReviewItemResponse(SalesReviewItemBase):
    id: int
    sales_review_id: int
    line_no: int
    amount: Decimal


class SalesReviewDeptOpinionResponse(BaseSchema):
    id: int
    sales_review_id: int
    review_round: int
    dept_code: str
    result: str
    opinion: Optional[str] = None
    reviewed_by: Optional[int] = None
    reviewed_by_name: Optional[str] = None
    reviewed_at: Optional[datetime] = None


class SalesReviewDeptOpinionSubmit(BaseSchema):
    result: str = Field(..., description="pass 或 fail")
    opinion: Optional[str] = Field(None, description="评审意见；fail 时建议填写原因")
    reviewed_by: Optional[int] = Field(
        None, description="评审人用户ID；缺省为当前操作人"
    )

    @model_validator(mode="after")
    def _validate_result(self):
        r = (self.result or "").strip().lower()
        if r not in ("pass", "fail"):
            raise ValueError("评审结果须为 pass 或 fail")
        # BaseSchema.validate_assignment=True：禁止 self.result= 触发二次校验递归
        object.__setattr__(self, "result", r)
        if r == "fail" and not (self.opinion or "").strip():
            raise ValueError("不通过时须填写评审意见")
        return self


class SalesReviewRejectRequest(BaseSchema):
    reason: Optional[str] = Field(None, description="驳回原因")


class SalesReviewBase(BaseSchema):
    customer_id: int = Field(..., description="客户ID")
    customer_code: Optional[str] = Field(None, max_length=50)
    customer_name: str = Field(..., max_length=200)
    customer_contact: Optional[str] = Field(None, max_length=100)
    customer_phone: Optional[str] = Field(None, max_length=50)
    project_name: str = Field(..., max_length=200, description="项目名称")
    review_date: Optional[date] = None
    delivery_date: Optional[date] = None
    urgency: str = Field("normal", max_length=20)
    risk_level: str = Field("medium", max_length=20)
    settlement_method: Optional[str] = Field(None, max_length=100)
    payment_cycle: Optional[str] = Field(None, max_length=100)
    delivery_location: Optional[str] = Field(None, max_length=200)
    transport_method: Optional[str] = Field(None, max_length=100)
    material_desc: Optional[str] = None
    spec_desc: Optional[str] = None
    process_desc: Optional[str] = None
    packaging_req: Optional[str] = None
    production_notes: Optional[str] = None
    sales_opinion: Optional[str] = None
    final_conclusion: Optional[str] = None
    remarks: Optional[str] = None
    attachments: Optional[List[Any]] = None
    quotation_id: Optional[int] = None
    quotation_code: Optional[str] = None
    customer_follow_up_id: Optional[int] = None
    salesman_id: Optional[int] = None
    salesman_name: Optional[str] = None


class SalesReviewCreate(SalesReviewBase):
    review_code: Optional[str] = Field(None, description="手工单号；空则按规则生成")
    items: List[SalesReviewItemCreate] = Field(default_factory=list)

    @model_validator(mode="after")
    def _require_items(self):
        if not self.items:
            raise ValueError("评审单必须至少包含一条明细")
        return self


class SalesReviewUpdate(BaseSchema):
    customer_id: Optional[int] = None
    customer_code: Optional[str] = None
    customer_name: Optional[str] = None
    customer_contact: Optional[str] = None
    customer_phone: Optional[str] = None
    project_name: Optional[str] = None
    review_date: Optional[date] = None
    delivery_date: Optional[date] = None
    urgency: Optional[str] = None
    risk_level: Optional[str] = None
    settlement_method: Optional[str] = None
    payment_cycle: Optional[str] = None
    delivery_location: Optional[str] = None
    transport_method: Optional[str] = None
    material_desc: Optional[str] = None
    spec_desc: Optional[str] = None
    process_desc: Optional[str] = None
    packaging_req: Optional[str] = None
    production_notes: Optional[str] = None
    sales_opinion: Optional[str] = None
    final_conclusion: Optional[str] = None
    remarks: Optional[str] = None
    attachments: Optional[List[Any]] = None
    quotation_id: Optional[int] = None
    quotation_code: Optional[str] = None
    customer_follow_up_id: Optional[int] = None
    salesman_id: Optional[int] = None
    salesman_name: Optional[str] = None
    items: Optional[List[SalesReviewItemCreate]] = None


class SalesReviewResponse(SalesReviewBase):
    id: int
    uuid: str
    tenant_id: int
    review_code: str
    status: str
    review_round: int
    sales_order_id: Optional[int] = None
    sales_order_code: Optional[str] = None
    total_quantity: Decimal = Decimal("0")
    total_amount: Decimal = Decimal("0")
    items: List[SalesReviewItemResponse] = Field(default_factory=list)
    dept_opinions: List[SalesReviewDeptOpinionResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime
    created_by: Optional[int] = None
    updated_by: Optional[int] = None
    created_by_name: Optional[str] = None
    updated_by_name: Optional[str] = None


class SalesReviewListItem(BaseSchema):
    id: int
    review_code: str
    customer_id: int
    customer_name: str
    project_name: str
    status: str
    review_round: int
    urgency: str
    risk_level: str
    delivery_date: Optional[date] = None
    review_date: Optional[date] = None
    total_quantity: Decimal = Decimal("0")
    total_amount: Decimal = Decimal("0")
    salesman_name: Optional[str] = None
    sales_order_code: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    created_by: Optional[int] = None
    created_by_name: Optional[str] = None
    updated_by: Optional[int] = None
    updated_by_name: Optional[str] = None


class SalesReviewListEnvelope(BaseSchema):
    items: List[SalesReviewListItem]
    total: int
    skip: int = 0
    limit: int = 50


class SalesReviewPushPreview(BaseSchema):
    can_push: bool
    blocking_reason: Optional[str] = None
    review_code: str
    customer_name: str
    item_count: int
    total_quantity: Decimal
    total_amount: Decimal
    items: List[Dict[str, Any]] = Field(default_factory=list)


class SalesReviewPushResult(BaseSchema):
    success: bool
    message: str
    sales_order_id: Optional[int] = None
    sales_order_code: Optional[str] = None
